gusto/forcing.py

```diff
@@ -26,106 +28,132 @@ def __init__(self, state, euler_poincare=True, linear=False, extra_terms=None):
             warning('Setting euler_poincare to False because you have set linear=True')
         else:
             self.euler_poincare = euler_poincare
+
+        # set up functions
+        self.Vu = state.spaces("HDiv")
+        self.x0 = Function(state.W)   # copy x to here
+        self.test = TestFunction(self.Vu)
+        self.trial = TrialFunction(self.Vu)
+        self.uF = Function(self.Vu)
+
+        # find out which terms we need
+        self.extruded = self.Vu.extruded
+        self.coriolis = state.Omega is not None or hasattr(state.fields, "coriolis")
+        self.sponge = state.mu is not None
+        self.topography = hasattr(state.fields, "topography")
         self.extra_terms = extra_terms
-        self._build_forcing_solver()
+
+        # some constants to use for scaling terms
+        self.scaling = Constant(1.)
+        self.mu_scaling = Constant(1.)
+
+        self._build_forcing_solvers()
+
+    def mass_term(self):
+        return inner(self.test, self.trial)*dx
+
+    def coriolis_term(self):
+        u0 = split(self.x0)[0]
+        return -inner(self.test, cross(2*self.state.Omega, u0))*dx
+
+    def sponge_term(self):
+        u0 = split(self.x0)[0]
+        return self.state.mu*inner(self.test, self.state.k)*inner(u0, self.state.k)*dx
+
+    def euler_poincare_term(self):
+        u0 = split(self.x0)[0]
+        return -0.5*div(self.test)*inner(u0, u0)*dx
 
     @abstractmethod
-    def _build_forcing_solver(self):
+    def pressure_gradient_term(self):
         pass
 
-    @abstractmethod
-    def apply(self, scale, x, x_nl, x_out, **kwargs):
+    def forcing_term(self):
+        L = self.pressure_gradient_term()
+        if self.extruded:
+            L += self.gravity_term()
+        if self.coriolis:
+            L += self.coriolis_term()
+        if self.euler_poincare:
+            L += self.euler_poincare_term()
+        if self.topography:
+            L += self.topography_term()
+        if self.extra_terms is not None:
+            L += inner(self.test, self.extra_terms)*dx
+        # scale L
+        L = self.scaling * L
```

Comment:
The scaling? This is what allows us to do half a step of explicit then half a step of implicit (or whatever) - the scaling constant is set in the apply method.